agent/proxycfg/state_test.go

```diff
@@ -112,3 +117,344 @@ func TestStateChanged(t *testing.T) {
+				Address: "10.0.1.1",
+				Port:    443,
+				Proxy: structs.ConnectProxyConfig{
+					DestinationServiceName: "web",
+					Upstreams: structs.Upstreams{
+						structs.Upstream{
+							DestinationType: structs.UpstreamDestTypePreparedQuery,
+							DestinationName: "query",
+							LocalBindPort:   10001,
+						},
+						structs.Upstream{
+							DestinationType: structs.UpstreamDestTypeService,
+							DestinationName: "api",
+							LocalBindPort:   10002,
+						},
+						structs.Upstream{
+							DestinationType: structs.UpstreamDestTypeService,
+							DestinationName: "api-failover-remote",
+							Datacenter:      "dc2",
+							LocalBindPort:   10003,
+							MeshGateway: structs.MeshGatewayConfig{
+								Mode: structs.MeshGatewayModeRemote,
+							},
+						},
+						structs.Upstream{
+							DestinationType: structs.UpstreamDestTypeService,
+							DestinationName: "api-failover-local",
+							Datacenter:      "dc2",
+							LocalBindPort:   10004,
+							MeshGateway: structs.MeshGatewayConfig{
+								Mode: structs.MeshGatewayModeLocal,
+							},
+						},
+						structs.Upstream{
+							DestinationType: structs.UpstreamDestTypeService,
+							DestinationName: "api-failover-direct",
+							Datacenter:      "dc2",
+							LocalBindPort:   10005,
+							MeshGateway: structs.MeshGatewayConfig{
+								Mode: structs.MeshGatewayModeNone,
+							},
+						},
+					},
+				},
+			},
+			sourceDC: "dc1",
+			stages: []verificationStage{
+				verificationStage{
+					requiredWatches: map[string]verifyWatchRequest{
+						rootsWatchID:                    genVerifyRootsWatch("dc1"),
+						leafWatchID:                     genVerifyLeafWatch("web", "dc1"),
+						intentionsWatchID:               genVerifyIntentionWatch("web", "dc1"),
+						"upstream:prepared_query:query": genVerifyPreparedQueryWatch("query", "dc1"),
+						"discovery-chain:api":           genVerifyDiscoveryChainWatch("api", "dc1"),
+						"upstream:" + serviceIDPrefix + "api-failover-remote?dc=dc2": genVerifyGatewayWatch("dc2"),
+						"upstream:" + serviceIDPrefix + "api-failover-local?dc=dc2":  genVerifyGatewayWatch("dc1"),
+						"upstream:" + serviceIDPrefix + "api-failover-direct?dc=dc2": genVerifyServiceWatch("api-failover-direct", "", "dc2", true),
+					},
+				},
+			},
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			state, err := newState(&tc.ns, "")
+
+			// verify building the initial state worked
+			require.NoError(t, err)
+			require.NotNil(t, state)
+
+			// setup the test logger to use the t.Log
+			state.logger = testutil.TestLogger(t)
+
+			// setup a new testing cache notifier
+			cn := newTestCacheNotifier()
+			state.cache = cn
+
+			// setup the local datacenter information
+			state.source = &structs.QuerySource{
+				Datacenter: tc.sourceDC,
+			}
+
+			// setup the ctx as initWatches expects this to be there
+			state.ctx, state.cancel = context.WithCancel(context.Background())
```

Comment:
Do you need to `defer state.cancel()` here as a backup in case the normal flow fails abruptly?

Reply:
Nothing is running async here. Some things expect the context to be there even if they don't care about them. So I didn't bother deferring anything.